### DISCUSSION TITLE
TP8 - EJ1 - ¿Puntero simple o puntero doble?

### DISCUSSION THREAD

**Discussion by Nikzza** [INGCOM-UNRN-P1/cursada-2023#200] [Preguntas y respuestas] [answered]
martin tengo una duda en la funcion de insertar_al_principio 
diste como funcion

void insertar_al_principio(nodo_t* lista, int valor)

lo intente hacer y no puedo hacer que se encadene con el siguiente "vagon"

lo unico que se me ocurrio es utilizar casi lo mismo pero con nodo_t** lista

estaria correcto?

sino no podrias acceder a la cabeza de la lista, porque lo que modifiques en la funcion moriria en la funcion

**Comment by IoachimusRoderici** [comment]
Según entiendo lo que hay que hacer es crear un nodo nuevo, conectarlo con el primer nodo de la lista (osea `nodo_nuevo.siguiente = primer_nodo_de_la_lista`), y devolver un pointer al nodo nuevo.

> porque lo que modifiques en la función moriría en la función.

Para que el nodo nuevo siga existiendo después de que termine la función, lo tenés que crear con `malloc`.

**Reply by Nikzza** [reply]
tal vez estoy equivocado pero la struct tiene que ser 
![image](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/141664052/1784d663-6ded-41e0-8b1e-d51912f40334)

necesitas un doble puntero para poder acceder a la lista osea:
nuevo_nodo->siguiente = *lista

**Reply by IoachimusRoderici** [reply]
`nuevo_nodo->siguiente = *lista`

La flecha está bien pero el asterisco está de más.

`lista` es de tipo `struct nodo *`, igual que `siguiente`, así que no hace falta el asterisco.

**Reply by Nikzza** [reply]
Exacto, pero para modificar la dirección de memoria de la lista hace falta inicializarla como

```C
nodo_t* lista = NULL;
```

y cuando la mandas a una función tiene que ser `función(&lista)`

**Reply by Nikzza** [reply]
cuando la mandes por parametro a una funcion en la que quieras modificar esa lista los argumentos de la funcion tiene que haber un (nodo_t** lista)

**Reply by IoachimusRoderici** [reply]
> y cuando la mandas a una funcion tiene que ser `funcion(&lista)`.

no, porque `lista` está declarada como `nodo_t*`, ya es un pointer. Así que cuando haces `funcion(lista)` estás pasando una referencia, no una copia.

**Reply by Nikzza** [reply]
dejame, probar vengo haciendo la mayoria con doble pointer asi q ahi veo si lo puedo cambiar entonces

**Reply by Nikzza** [reply]
deberia inicializar entonces struct nodo siguiente nada mas? basicamente no seria un puntero

**Comment by martinvilu** [comment]
¿Por qué doble puntero? Es correcto pensar en uno, pero la razón es más importante todavía.

Teniendo en cuenta que el ejercicio 2 no lo necesita.

**Reply by Nikzza** [reply]
pasa que si utilizamos un doble puntero la funcion tendria que retornar un la lista nueva, sino los cambios no se podrian hacer

**Reply by Nikzza** [reply]
Con un puntero doble, podes modificar la estructura original y reflejar esos cambios en la variable fuera de la función. Pero si solo pasas un puntero , la función opera en una copia del puntero y los cambios no se reflejan

**Reply by alengeor** [reply]
Buenas grupo, profe @martinvilu . A mi también me llamo la atención que las funciones estén declaradas como tipo void y que reciban como argumento el puntero simple. En la confusión, investigué por mi cuenta y no encontré un ejemplo que vincule ambas características: o era de tipo void con puntero doble, o retornaba una estructura de tipo nodo_t recibiendo puntero simple como parámetro. Decidí realizar las funciones con doble puntero, pasando la direccion de la lista (&lista como parametro), manteniendo el tipo void. Recién ahora terminé el primer punto, con la gestión de archivos, luego de dedicarle gran parte de la tarde. Solo me queda documentar el punto 1 y encarar el punto 2.

**Reply by martinvilu** [reply]
Pero por que, buscar en internet y ver que esta resuelto de esa forma no alcanza

**Reply by martinvilu** [reply]
> Con un puntero doble, podes modificar la estructura original y reflejar esos cambios en la variable fuera de la función. Pero si solo pasas un puntero , la función opera en una copia del puntero y los cambios no se reflejan

¿Estas seguro de que es lo que estas modificando?

**Reply by martinvilu** [reply]
> Para que el nodo nuevo siga existiendo después de que termine la función, lo tenés que crear con `malloc`.

No estrictamente, pero como está planteado sí.

**Reply by alengeor** [reply]
Ahora entiendo la pregunta. Como vimos, por defecto los parametros a funcion en C se pasan por copia y cualquier cambio no afecta la variable original. Si usas nodo_t* lista estás pasando una copia, lo que hace que no se modifique la lista original. En cambio, si pasamos puntero de puntero (nodo_t** lista) estamos pasando una referencia a la lista original (nodo_t* lista), permitiendo que la función realice cambios en la lista original. 
La otra alternativa era que retorne dato de tipo nodo_t, entonces la "actualizacion" de la lista se realiza donde se llamo a la funcion, jugando con el retorno de esta.

**Reply by martinvilu** [reply]
Estás muy cerca, pero pensá en que pasa cuando la lista está vacía.

De la presentación, de qué tipo es 'inicio'

![imagen](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/56625/d3ebfa7b-78ad-46f8-abae-51c3c2a7fbde)

Y mira como es una lista con un solo nodo.

![imagen](https://github.com/INGCOM-UNRN-P1/cursada-2023/assets/56625/9694e699-320d-4e8e-90c8-8d0d1ee14d19)

¿Cómo quedaría gráficamente la propuesta de @Nikzza ?

**Reply by Nikzza** [reply]
quedaria como CABEZA -> un_nodo > [sgt][1] -> NULL

**Reply by martinvilu** [reply]
Cómo puntero doble, no tiene "nombre propio"

**Reply by Nikzza** [reply]
al recorrer una lista por ejemplo no hace falta un doble puntero, porque solamente es recorrer y no tenes que modificar el puntero que apunta cabeza, en cambio, en insertar un elemento al principio de la lista, debes acceder a cabeza para que apunte al nodo que se insertó ahora, es lo que me explicaste cuando te pregunte sobre el g1 de pileta, no es tan solo insertar un nodo al principio y ya esta, cabeza se deberia modificar y apuntar al primer nodo.

**Reply by Nikzza** [reply]
basicamente como lo podria explicar yo seria:
En un caso de insercion al principio, con una funcion tipo void:

cabeza es un puntero que apunta al puntero original que apunta al primer nodo de la lista
Con cabeza, podes acceder y modificar el puntero original, entonces el puntero original se actualiza para que apunte al nuevo nodo que insertaste al principio de la lista. 
Sin un puntero a puntero, no podrías modificar el puntero original y, entonces, no podrías realizar la inserción al principio 

En un caso de insercion al principio, con una funcion de tipo struct nodo:

En este caso no hace falta utilizar un puntero a puntero. Porque?
porque los cambios que estas realizando en la funcion los podes retornar y esos valores modificados se le atribuyen a una variable de tipo nodo_t

**Reply by martinvilu** [reply]
Lo más importante de todo esto
> Sin un puntero a puntero, no podrías modificar el puntero original y, entonces, no podrías realizar la inserción al principio

Fantástico.

Con un puntero simple funciona también, pero el uso es ligeramente distinto.

Ahora, compara con la estructura del ejercicio dos.

**Reply by Nikzza** [reply]
Me podrias mostrar como funcionaria con un puntero simple para poder modificar la lista? sinceramente no encontre la vuelta, con un ejemplo basico me sobra para entenderlo

**Reply by martinvilu** [reply, marked as answer]
En el único caso en el que no funciona (directamente) el puntero simple, es cuando pasas un puntero `NULL`, que es cuando la lista está vacía y las operaciones que cambian el `inicio`.

Simplemente, por no poder modificar el puntero original.

**Reply by Nikzza** [reply]
Ahhhhh, ahora si tiene sentido, gracias

**Reply by Nikzza** [reply]
vas a tener que utilizar un puntero doble para modificar el puntero original cuando haya algo que se quiera insertar al principio, es que basicamente vas a tener que utilizarlo para la mayoria de las funciones que te piden modificar. Ya que, si se quiere insertar al final y la lista esta vacia vas a tener que acceder al puntero original para poder modificarlo y que apunte al primer nodo

**Reply by Nikzza** [reply]
lo mismo con insertar en posicion, si la posicion es 0, osea el principio, deberias hacer lo mismo
